proposals/2022-09-14-kubernetes-versioning.md

```diff
@@ -0,0 +1,208 @@
+# Wasm Shim Versioning in Kubernetes
+Wasm workloads in Kubernetes will likely need to lock into a specific version of a Wasm runtime. This proposal is intended to address this problem.
+
+
+## <a name='TableofContents'></a>Table of Contents
+
+<!-- vscode-markdown-toc -->
+* [Table of Contents](#TableofContents)
+* [Background](#Background)
+* [Problem Statement](#ProblemStatement)
+	* [Goals](#Goals)
+	* [Non Goals](#NonGoals)
+* [Proposal](#Proposal)
+	* [Versioning of both shim binary and RuntimeClass name](#VersioningofbothshimbinaryandRuntimeClassname)
+	* [Introducing a new version of a shim](#Introducinganewversionofashim)
+* [Alternatives](#Alternatives)
+	* [Alternative #1](#Alternative1)
+		* [Pros](#Pros)
+		* [Cons](#Cons)
+		* [Conclusion](#Conclusion)
+* [Additional Details](#AdditionalDetails)
+	* [Test Plan](#TestPlan)
+* [Implementation History](#ImplementationHistory)
+
+<!-- vscode-markdown-toc-config
+	numbering=false
+	autoSave=false
+	/vscode-markdown-toc-config -->
+<!-- /vscode-markdown-toc -->
+
+## <a name='Background'></a>Background
+
+It is likely behaviors in a shim may change over time as new features are introduced or as existing defects are resolved. Pod workloads that target a version N of a shim may not be able to execute on version N+1. This proposal describes a solution to ensure Pod workloads targeting a specific version of a shim will always be scheduled on a node that contains a safe version of a shim for the workload.
+
+## <a name='ProblemStatement'></a>Problem Statement
+The containerd shims in this repository are installed on Kubernetes nodes. The containerd config on each Kubernetes node has a handler registered for each shim, a handler name mapped to binary on the PATH. A RuntimeClass is registered in Kubernetes describing the containerd handler needed to run a Pod workload, as well as label selectors to inform the Kubernetes scheduler to only schedule workloads on particular nodes.
+
+1. A shim binary is installed on a Kubernetes node at `/bin/containerd-shim-spin-v1`.
+2. The following shows the containerd config.toml that maps the runtime type to the shim binary from step 1.
+      ```toml
+        [plugins.cri.containerd.runtimes.spin]
+          runtime_type = "io.containerd.spin.v1"
```

Comment:
I assume affinity between the binary @ `/bin/containerd-shim-spin-v1` and the `config.toml` configuration above is that containerd plugins by default map the `runtime_type` substring value after the `io.containerd.` string suffix to the filepath @ `/bin/containerd-shim-<insert runtime_type suffix here and substitute "." with "_"`. If there's a more verbose way to express this (like a `handler_runtime_path` config key or something) that might be appropriate for this example to make the affinity clearer.

Reply:
We might be able to do something here, but I think it's constrained: https://github.com/containerd/containerd/blob/main/docs/cri/config.md#runtime-classes